js/src/forum/states/ComposerState.js

```diff
@@ -0,0 +1,341 @@
+  }
+
+  /**
+   * Show the composer.
+   *
+   * @public
+   */
+  show() {
+    if (this.position === ComposerState.Position.NORMAL || this.position === ComposerState.Position.FULLSCREEN) return;
+
+    this.position = ComposerState.Position.NORMAL;
+    m.redraw();
+  }
+
+  /**
+   * Close the composer.
+   *
+   * @public
+   */
+  hide() {
+    this.clear();
+    m.redraw();
+  }
+
+  /**
+   * Confirm with the user so they don't lose their content, then close the
+   * composer.
+   *
+   * @public
+   */
+  close() {
+    if (!this.preventExit()) {
+      this.hide();
+    }
+  }
+
+  /**
+   * Minimize the composer. Has no effect if the composer is hidden.
+   *
+   * @public
+   */
+  minimize() {
+    if (!this.isVisible()) return;
+
+    this.position = ComposerState.Position.MINIMIZED;
+    m.redraw();
+  }
+
+  /**
+   * Take the composer into fullscreen mode. Has no effect if the composer is
+   * hidden.
+   *
+   * @public
+   */
+  fullScreen() {
+    if (!this.isVisible()) return;
+
+    this.position = ComposerState.Position.FULLSCREEN;
+    m.redraw();
+  }
+
+  /**
+   * Exit fullscreen mode.
+   *
+   * @public
+   */
+  exitFullScreen() {
+    if (this.position !== ComposerState.Position.FULLSCREEN) return;
+
+    this.position = ComposerState.Position.NORMAL;
+    m.redraw();
+  }
+
+  /**
+   * Determine whether the body matches the given component class and data.
+   *
+   * @param {object} type The component class to check against. Subclasses are
+   *                      accepted as well.
+   * @param {object} data
+   * @return {boolean}
+   */
+  bodyMatches(type, data = {}) {
+    // Fail early when the body is of a different type
+    if (!subclassOf(this.body.componentClass, type)) return false;
+
+    // Now that the type is known to be correct, we loop through the provided
+    // data to see whether it matches the data in the attributes for the body.
+    return Object.keys(data).every((key) => this.body.attrs[key] === data[key]);
+  }
+
+  /**
+   * Determine whether or not the Composer is visible.
+   *
+   * True when the composer is displayed on the screen and has a body component.
+   * It could be open in "normal" or full-screen mode, or even minimized.
+   *
+   * @returns {boolean}
+   */
+  isVisible() {
+    return this.position !== ComposerState.Position.HIDDEN;
+  }
+
+  /**
+   * Determine whether or not the Composer is covering the screen.
+   *
+   * This will be true if the Composer is in full-screen mode on desktop,
+   * or if we are on a mobile device, where we always consider the composer as full-screen..
+   *
+   * @return {Boolean}
+   * @public
+   */
+  isFullScreen() {
+    return this.position === ComposerState.Position.FULLSCREEN || this.onMobile();
+  }
+
+  /**
+   * Determine whether we are on mobile.
+   *
+   * @return {Boolean}
+   * @public
+   */
+  onMobile() {
+    // 767 is the mobile screen cutoff defined in the less variables file
+    return window.innerWidth <= 767;
```

Comment:
It's possible to expose the variable in less:
```css
:root {
 --phone: @phone;
}   
```

I think it'd be also better to move that method into global state, since it's something other components might use aswell. 

Inside ForumApplication mount:
```js
const styles = getComputedStyle(document.documentElement)
const phone = styles.getPropertyValue('--phone')

this.isMobile = function () {
   return window.matchMedia(phone).matches
}
```

Reply:
That's a really good point! This particular code is actually something we weren't happy with during internal review, and I like what you're proposing a lot more! @franzliedke how would you feel about this?

Reply:
I forgot about CSS variables when we discussed this! They should be supported by every browser that supports our JS syntax (see https://caniuse.com/#feat=css-variables).

Reply:
@w-4 Great idea, thank you very much! And :+1: for daring to venture into this code review, we appreciate it.

I've applied this idea, with a twist: instead of exposing and then matching the media query, I now expose a CSS variable that is changed in CSS based on the media query and returns the mode. Also added this to the `Application` base class.